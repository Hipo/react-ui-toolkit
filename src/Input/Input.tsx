import React from "react";
import classNames from "classnames";
import styled from "styled-components";

const InputContainer = styled.div`
  display: flex;
`;
const PrefixIcon = styled.span`
  background: white;

  border-right: 1px solid grey;
`;
const SuffixIcon = styled.span`
  background: white;

  border-left: 1px solid grey;
`;

interface InputProps {
  name: string;
  type?: string;
  value?: string;
  onChange: React.ReactEventHandler<HTMLInputElement>;
  onFocus?: React.ReactEventHandler<HTMLInputElement>;
  onBlur?: React.ReactEventHandler<HTMLInputElement>;
  onKeyDown?: React.KeyboardEventHandler<HTMLInputElement>;
  onKeyUp?: React.KeyboardEventHandler<HTMLInputElement>;
  placeholder?: string;
  min?: number;
  max?: number;
  step?: number;
  role?: string;
  prefixIcon?: React.ReactNode;
  suffixIcon?: React.ReactNode;
  isDisabled?: boolean;
  className?: string;
  inputContainerRef?: React.RefObject<HTMLDivElement>;
}

function Input(props: InputProps) {
  const {
    name,
    type = "text",
    value,
    onChange,
    placeholder,
    step,
    min,
    max,
    isDisabled,
    onFocus,
    className,
    onBlur,
    onKeyUp,
    onKeyDown,
    prefixIcon,
    suffixIcon,
    role,
    inputContainerRef,
    ...rest
  } = props;
  const inputContainerClassName = classNames("input-container", className);
  const inputClassName = classNames("input", {
    disabled: isDisabled
  });

  return (
    <InputContainer
      ref={inputContainerRef}
      role={role}
      className={inputContainerClassName}>
      {prefixIcon && (
        <PrefixIcon className={"input-container-prefix-icon"}>{prefixIcon}</PrefixIcon>
      )}

      <input
        className={inputClassName}
        name={name}
        type={type}
        step={step}
        max={max}
        min={min}
        value={value}
        onChange={onChange}
        onFocus={onFocus}
        onBlur={onBlur}
        placeholder={placeholder}
        autoComplete={"off"}
        autoCorrect={"off"}
        disabled={isDisabled}
        onKeyDown={onKeyDown}
        onKeyUp={onKeyUp}
        {...rest}
      />

      {suffixIcon && (
        <SuffixIcon className={"input-container-suffix-icon"}>{suffixIcon}</SuffixIcon>
      )}
    </InputContainer>
  );
}

export default Input;
