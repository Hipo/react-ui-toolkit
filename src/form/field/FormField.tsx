import "./_form-field.scss";

import React from "react";
import classNames from "classnames";

import FormFieldMessage from "./message/FormFieldMessage";
import List from "../../list/List";
import ListItem from "../../list/item/ListItem";

export interface FormFieldProps {
  children: React.ReactNode;
  label?: string;
  labelledBy?: string;
  labelFor?: string;
  customClassName?: string;
  helperMessages?: string[];
  errorMessages?: string[];
  testid?: string;
}

function FormField(props: FormFieldProps) {
  const {
    label,
    labelledBy,
    labelFor,
    customClassName,
    children,
    errorMessages,
    helperMessages,
    testid
  } = props;
  const hasErrorMessage = Boolean(errorMessages?.length);
  const hasHelperMessage = Boolean(helperMessages?.length);
  const formFieldClassName = classNames("form-field", customClassName, {
    "form-field--has-error": hasErrorMessage
  });

  return (
    <div className={formFieldClassName} data-testid={testid}>
      <label id={labelledBy} htmlFor={labelFor} className={"form-field__label"}>
        {Boolean(label) && <span className={"form-field__title"}>{label}</span>}

        {children}
      </label>

      {hasErrorMessage && (
        <List
          testid={`${testid}.error-messages`}
          customClassName={"form-field__error-message-list"}
          items={errorMessages!}
        >
          {(message, messageTestId) => (
            <ListItem customClassName={"form-field__error-message-list__item"}>
              <FormFieldMessage type={"error"} message={message} testid={messageTestId} />
            </ListItem>
          )}
        </List>
      )}

      {!hasErrorMessage && hasHelperMessage && (
        <List
          testid={`${testid}.helper-messages`}
          customClassName={"form-field__helper-message-list"}
          items={helperMessages!}
        >
          {(message, messageTestId) => (
            <ListItem customClassName={"form-field__helper-message-list__item"}>
              <FormFieldMessage
                type={"helper"}
                message={message}
                testid={messageTestId}
              />
            </ListItem>
          )}
        </List>
      )}
    </div>
  );
}

export default FormField;
