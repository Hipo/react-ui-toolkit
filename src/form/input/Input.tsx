import "./_input.scss";

import React, {useState, useEffect} from "react";
import classNames from "classnames";

import {
  getNumberSeparators,
  parseNumber,
  mapDigitsToLocalVersion,
  formatNumber,
  removeLeadingZeros,
  getNegativeZero,
  getThousandthSeparatorCount
} from "../../core/utils/number/numberUtils";
import {getInputLocalizationOptions, getInputParseNumberOptions} from "./util/inputUtils";
import {InputProps} from "./util/inputTypes";

const Input = React.forwardRef<HTMLInputElement, InputProps>(
  /* eslint-disable complexity */
  (props, ref) => {
    const {
      testid,
      value,
      type = "text",
      isDisabled,
      hasError,
      customClassName,
      leftIcon,
      rightIcon,
      localizationOptions = {},
      role,
      autoComplete = "off",
      autoCorrect = "off",
      onChange,
      ...rest
    } = props;
    const {
      shouldFormatToLocaleString = false,
      locale,
      maximumFractionDigits = 0
    } = getInputLocalizationOptions(localizationOptions);
    const [
      {
        DECIMAL_NUMBER_SEPARATOR: decimalSeparatorForLocale,
        MINUS_SIGN: minusSignForLocale,
        LOCALE_NEGATIVE_ZERO: negativeZeroForLocale
      },
      setNumberSeparatorsForLocale
    ] = useState(() => ({
      ...getNumberSeparators(locale),
      ...getNegativeZero(locale)
    }));
    const isNumberInput = type === "number";
    const inputContainerClassName = classNames(
      "input-container",
      customClassName,
      `input-container--type-${type}`
    );
    const inputClassName = classNames("input", {
      "input-container--is-disabled": isDisabled,
      "input-container--has-error": hasError
    });
    let finalValue = value;

    if (
      !(
        typeof maximumFractionDigits === "number" &&
        Number.isInteger(maximumFractionDigits) &&
        maximumFractionDigits >= 0
      )
    ) {
      throw new Error("maximumFractionDigits should be zero or a positive integer.");
    }

    if (isNumberInput && typeof value === "string" && shouldFormatToLocaleString) {
      const [integerPart, decimalPart] = String(value).split(".");
      const numberFormatter = formatNumber({
        maximumFractionDigits,
        locale
      });

      // IF there is a decimal part or the value ends with ".",
      // make sure we add the decimal separator and map each digit on the decimal part to localized versions.
      // We shouldn't use parseInt or parseFloat with numberFormat util here because that removes zeros on the decimal part and disallows users to write something like: 10.01 or 10.102
      if (value.match(/\.$/)?.length || decimalPart) {
        finalValue = `${numberFormatter(
          parseInt(integerPart)
        )}${decimalSeparatorForLocale}${mapDigitsToLocalVersion({locale}, decimalPart)}`;
      } else if (integerPart) {
        if (integerPart !== minusSignForLocale && integerPart !== negativeZeroForLocale) {
          finalValue = numberFormatter(parseInt(integerPart));
        } else {
          finalValue = `${minusSignForLocale}${mapDigitsToLocalVersion(
            {locale},
            integerPart
          )}`;
        }
      }
    }

    useEffect(() => {
      setNumberSeparatorsForLocale({
        ...getNumberSeparators(locale),
        ...getNegativeZero(locale)
      });
    }, [locale]);

    return (
      <div role={role} className={inputContainerClassName} data-testid={testid}>
        {leftIcon && (
          <span className={"input-container__icon input-container__left-icon"}>
            {leftIcon}
          </span>
        )}

        <input
          ref={ref}
          className={inputClassName}
          type={isNumberInput ? "text" : type}
          autoComplete={autoComplete}
          value={finalValue}
          autoCorrect={autoCorrect}
          disabled={isDisabled}
          onChange={handleChange}
          {...rest}
        />

        {rightIcon && (
          <span className={"input-container__icon input-container__right-icon"}>
            {rightIcon}
          </span>
        )}
      </div>
    );

    function handleChange(event: React.SyntheticEvent<HTMLInputElement>) {
      if (isNumberInput) {
        const {value: newValue} = event.currentTarget;

        if (newValue) {
          const formattedNewValue = parseNumber(
            getInputParseNumberOptions({locale, maximumFractionDigits}),
            newValue
          );
          // Number("-") returns NaN. Should allow minus sign as first character.
          const isFormattedNewValueNotAValidNumber =
            formattedNewValue === "" ||
            (newValue !== "-" && Number.isNaN(Number(formattedNewValue)));
          let finalEventValue = formattedNewValue ? String(formattedNewValue) : "";

          // IF the parsed number is valid and there is a decimal separator,
          // we need to save the number as it is so that decimal part doesn't disappear
          if (
            !isFormattedNewValueNotAValidNumber &&
            formattedNewValue.match(/./)?.length
          ) {
            finalEventValue = String(formattedNewValue);
          } else if (isFormattedNewValueNotAValidNumber) {
            // IF the parsed number is not valid, we revert back to the valid value
            finalEventValue = value as string;
          }

          // IF 'shouldFormatToLocaleString' or 'maximumFractionDigits' are defined or the value is negative,
          // value can't have leading zeros. Like 0,000,123 or 010.50 or -00
          if (
            !isFormattedNewValueNotAValidNumber &&
            (shouldFormatToLocaleString ||
              maximumFractionDigits > 0 ||
              finalEventValue.includes("-"))
          ) {
            finalEventValue = removeLeadingZeros(locale, finalEventValue);
          }

          // IF maximumFractionDigits is set as 0, value can not be negative zero
          if (maximumFractionDigits === 0 && finalEventValue === "-0") {
            finalEventValue = "0";
          }

          // IF shouldFormatToLocaleString is defined, caret position should calculate according to thoudsandths separator count
          if (shouldFormatToLocaleString) {
            const thousandthsSeparatorCount = getThousandthSeparatorCount(
              finalEventValue
            );
            const prevValueThousandthsSeparatorCount = getThousandthSeparatorCount(
              value as string
            );
            const element = event.currentTarget;
            let caret = event.currentTarget.selectionStart || 0;

            if (prevValueThousandthsSeparatorCount === thousandthsSeparatorCount + 1) {
              caret -= 1;
            } else if (
              prevValueThousandthsSeparatorCount ===
              thousandthsSeparatorCount - 1
            ) {
              caret += 1;
            }

            window.requestAnimationFrame(() => {
              element.selectionStart = caret;
              element.selectionEnd = caret;
            });
          }

          event.currentTarget.value = finalEventValue;
        }
      }

      onChange(event);
    }
  }
  /* eslint-enable complexity */
);

export default Input;
