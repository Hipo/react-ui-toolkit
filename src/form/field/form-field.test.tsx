import React from "react";
import {render} from "@testing-library/react";
import "@testing-library/jest-dom";
import {create} from "react-test-renderer";

import {testA11y} from "../../core/utils/test/testUtils";
import FormField, {FormFieldProps} from "./FormField";
import Input from "../input/Input";

describe("<FormField />", () => {
  const defaultFormFieldProps: FormFieldProps = {
    testid: "form-field",
    children: <Input onChange={jest.fn} name={"test"} testid={"form-field.input"} />
  };

  it("should render correctly", () => {
    render(<FormField {...defaultFormFieldProps} />);
  });

  it("should match snapshot", () => {
    const tree = create(<FormField {...defaultFormFieldProps} />).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("should pass a11y test", async () => {
    const {container} = render(<FormField {...defaultFormFieldProps} />);

    await testA11y(container, {rules: {label: {enabled: false}}});
  });

  it("should render children correctly", () => {
    const {getByTestId} = render(<FormField {...defaultFormFieldProps} />);

    const childrenContent = getByTestId("form-field.input");

    expect(getByTestId(defaultFormFieldProps.testid!)).toContainElement(childrenContent);
  });

  it("should render label correctly", () => {
    const {getByTestId} = render(
      <FormField
        label={"Test"}
        labelledBy={"form-field.labelled-by"}
        labelFor={"form-field.input"}
        {...defaultFormFieldProps}
      />
    );

    const formFieldLabel = document.getElementsByTagName("label")[0];

    expect(getByTestId(defaultFormFieldProps.testid!)).toHaveTextContent("Test");
    expect(formFieldLabel).toHaveAttribute("id", "form-field.labelled-by");
    expect(formFieldLabel).toHaveAttribute("for", "form-field.input");
  });

  it("should handle errorMessages correctly", () => {
    const formFieldErrorMessages = ["Test Error 1", "Test Error 2"];
    const {getByTestId} = render(
      <FormField errorMessages={formFieldErrorMessages} {...defaultFormFieldProps} />
    );

    expect(getByTestId(defaultFormFieldProps.testid!)).toHaveClass(
      "form-field--has-error"
    );

    formFieldErrorMessages.forEach((errorMessage) => {
      expect(getByTestId(defaultFormFieldProps.testid!)).toHaveTextContent(errorMessage);
    });

    expect(
      getByTestId(`${defaultFormFieldProps.testid}.error-messages`).children
    ).toHaveLength(formFieldErrorMessages.length);
  });

  it("should handle helperMessages correctly", () => {
    const formFieldHelperMessages = ["Test Helper 1", "Test Helper 2"];
    const {getByTestId} = render(
      <FormField helperMessages={formFieldHelperMessages} {...defaultFormFieldProps} />
    );

    formFieldHelperMessages.forEach((helperMessage) => {
      expect(getByTestId(defaultFormFieldProps.testid!)).toHaveTextContent(helperMessage);
    });

    expect(
      getByTestId(`${defaultFormFieldProps.testid}.helper-messages`).children
    ).toHaveLength(formFieldHelperMessages.length);
  });

  it("should not render helper messages if error messages exists", () => {
    const formFieldErrorMessages = ["Test Error 1", "Test Error 2"];
    const formFieldHelperMessages = ["Test Helper 1", "Test Helper 2"];

    const {getByTestId} = render(
      <FormField
        helperMessages={formFieldHelperMessages}
        errorMessages={formFieldErrorMessages}
        {...defaultFormFieldProps}
      />
    );

    expect(getByTestId(defaultFormFieldProps.testid!)).toHaveClass(
      "form-field--has-error"
    );

    formFieldErrorMessages.forEach((errorMessage) => {
      expect(getByTestId(defaultFormFieldProps.testid!)).toHaveTextContent(errorMessage);
    });

    formFieldHelperMessages.forEach((helperMessage) => {
      expect(getByTestId(defaultFormFieldProps.testid!)).not.toHaveTextContent(
        helperMessage
      );
    });

    expect(
      getByTestId(`${defaultFormFieldProps.testid}.error-messages`).children
    ).toHaveLength(formFieldErrorMessages.length);

    const helperMessagesContent = document.getElementsByClassName(
      "form-field__helper-message-list"
    )[0];

    expect(helperMessagesContent).toBeUndefined();
  });
});
