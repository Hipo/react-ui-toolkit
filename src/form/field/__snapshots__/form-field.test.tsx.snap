// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormField /> should match snapshot 1`] = `
<div
  className="form-field"
  data-testid="form-field"
>
  <label
    className="form-field__label"
  >
    <div
      className="input-container input-container--type-text"
      data-testid="form-field.input"
    >
      <input
        autoComplete="off"
        autoCorrect="off"
        className="input"
        name="test"
        onChange={[Function]}
        type="text"
      />
    </div>
  </label>
</div>
`;
